broker-core/src/test/java/io/zeebe/broker/job/JobStateControllerTest.java

```diff
@@ -0,0 +1,263 @@
+  public void setUp() throws Exception {
+    stateController = new JobStateController();
+    stateController.open(folder.newFolder("db"), false);
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    stateController.close();
+  }
+
+  @Test
+  public void shouldCreateJobEntry() {
+    // given
+    final long key = 1L;
+    final JobRecord jobRecord = createJobRecord();
+
+    // when
+    stateController.create(key, jobRecord);
+
+    // then
+    assertThat(stateController.exists(key)).isTrue();
+    assertJobState(key, State.ACTIVATABLE);
+    assertJobRecordIsEqualTo(stateController.getJob(key), jobRecord);
+  }
+
+  @Test
+  public void shouldActivateJob() {
+    // given
+    final long key = 1L;
+    final JobRecord jobRecord = createJobRecord();
+
+    // when
+    stateController.activate(key, jobRecord);
+
+    // then
+    assertThat(stateController.exists(key)).isTrue();
+    assertJobState(key, State.ACTIVATED);
+    assertJobRecordIsEqualTo(stateController.getJob(key), jobRecord);
+  }
+
+  @Test
+  public void shouldTimeoutJob() {
+    // given
+    final long key = 1L;
+    final JobRecord jobRecord = createJobRecord();
+
+    // when
+    stateController.timeout(key, jobRecord);
+
+    // then
+    assertThat(stateController.exists(key)).isTrue();
+    assertJobState(key, State.ACTIVATABLE);
+    assertJobRecordIsEqualTo(stateController.getJob(key), jobRecord);
+  }
+
+  @Test
+  public void shouldDeleteJob() {
+    // given
+    final long key = 1L;
+    final JobRecord jobRecord = createJobRecord();
+
+    // when
+    stateController.create(key, jobRecord);
+
+    // then
+    assertThat(stateController.exists(key)).isTrue();
+
+    // when
+    stateController.delete(key, jobRecord);
+
+    // then
+    assertThat(stateController.exists(key)).isFalse();
+    assertThat(State.values()).noneMatch(state -> stateController.isInState(key, state));
+    assertThat(stateController.getJob(key)).isNull();
+  }
+
+  @Test
+  public void shouldFailJobWithRetriesLeft() {
+    // given
+    final long key = 1L;
+    final JobRecord jobRecord = createJobRecord().setRetries(1);
+
+    // when
+    stateController.fail(key, jobRecord);
+
+    // then
+    assertThat(stateController.exists(key)).isTrue();
+    assertJobState(key, State.ACTIVATABLE);
+    assertJobRecordIsEqualTo(stateController.getJob(key), jobRecord);
+  }
+
+  @Test
+  public void shouldFailJobWithNoRetriesLeft() {
+    // given
+    final long key = 1L;
+    final JobRecord jobRecord = createJobRecord().setRetries(0);
+
+    // when
+    stateController.fail(key, jobRecord);
+
+    // then
+    assertThat(stateController.exists(key)).isTrue();
+    assertJobState(key, State.FAILED);
+    assertJobRecordIsEqualTo(stateController.getJob(key), jobRecord);
+  }
+
+  @Test
+  public void shouldResolveJob() {
+    // given
+    final long key = 1L;
+    final JobRecord jobRecord = createJobRecord();
+
+    // when
+    stateController.fail(key, jobRecord);
```

Comment:
How it is resolved with fail ? :D

Reply:
Haha, this worked because it was failing with some retries left, which essentially does the same thing (makes it activatable again).